src/messages/mod.rs

```diff
@@ -37,62 +35,114 @@ use xor_name::{Prefix, XorName};
+    /// ask for a longer chain that can be recodnised). Therefor we don't need to sign this field.
+    src: SrcAuthority,
     /// The body of the message.
-    pub variant: Variant,
+    variant: Variant,
     /// Source's knowledge of the destination section key. If present, the destination can use it
     /// to determine the length of the proof of messages sent to the source so the source would
     /// trust it (the proof needs to start at this key).
-    pub dst_key: Option<bls::PublicKey>,
-}
-
-/// Partially deserialized message.
-#[derive(Clone, Eq, PartialEq, Hash, Serialize, Deserialize, Debug)]
-pub struct PartialMessage {
-    /// Destination location.
-    pub dst: DstLocation,
-}
+    dst_key: Option<bls::PublicKey>,
 
-impl PartialMessage {
-    /// Deserialize the message.
-    pub fn from_bytes(bytes: &Bytes) -> Result<Self> {
-        Ok(bincode::deserialize(&bytes[..])?)
-    }
+    /// Serialised message, this is a signed and fully serialised message ready to send.
+    #[serde(skip)]
+    serialized: Bytes,
+    #[serde(skip)]
+    hash: MessageHash,
 }
 
 impl Message {
-    /// Deserialize the message.
+    /// Deserialize the message. Only called on message receipt.
+    #[allow(unused)] // prefix field not used below
     pub(crate) fn from_bytes(bytes: &Bytes) -> Result<Self> {
-        Ok(bincode::deserialize(&bytes[..])?)
+        let mut msg: Message = bincode::deserialize(&bytes[..])?;
+        let signed_bytes = serialize_for_signing(&msg.dst, msg.dst_key.as_ref(), &msg.variant)?;
+        match msg.src.clone() {
+            SrcAuthority::Node {
+                public_id,
+                signature,
+            } => {
+                if public_id.verify(&signed_bytes, &signature) {
+                    msg.serialized = bytes.clone();
+                    msg.hash = MessageHash::from_bytes(bytes);
+                    Ok(msg)
+                } else {
+                    Err(RoutingError::FailedSignature)
+                }
+            }
+            SrcAuthority::Section {
+                prefix,
+                signature,
+                proof,
+            } => {
+                // FIXME Assumes the nodes proof last key is the one signing this message
+                if proof.last_key().verify(&signature, &signed_bytes) {
+                    msg.serialized = bytes.clone();
+                    msg.hash = MessageHash::from_bytes(bytes);
+                    Ok(msg)
+                } else {
+                    Err(RoutingError::FailedSignature)
+                }
+            }
+        }
     }
 
-    /// Serialize the message.
-    pub(crate) fn to_bytes(&self) -> Result<Bytes> {
-        Ok(bincode::serialize(self)?.into())
+    /// send across wire
+    pub(crate) fn to_bytes(&self) -> Bytes {
+        self.serialized.clone()
+    }
+
+    /// Creates a signed message where signature is assumed valid.
+    pub fn new_signed(
```

Comment:
This probably should't be `pub`, because it can be used to violate the invariant (that the signature is always valid).